File: android/app/src/main/java/com/tanay/echo/snippet/Snippets.kt
package com.tanay.echo.snippet

/** A user-defined voice snippet: speak [cue], and Echo pastes [expansion] (a canned block). */
data class Snippet(val cue: String, val expansion: String)

/**
 * If [text] (what you just dictated) matches a snippet's [Snippet.cue] — ignoring case, surrounding
 * whitespace, collapsed inner spaces, and trailing sentence punctuation — return its expansion; else
 * null. The match is on the whole utterance, so saying a cue inside a longer sentence inserts the
 * words literally (near-zero false positives). First match wins.
 */
fun expandSnippet(text: String, snippets: List<Snippet>): String? {
    val key = normalizeCue(text)
    if (key.isEmpty()) return null
    return snippets.firstOrNull { normalizeCue(it.cue) == key }?.expansion
}

fun filterSnippets(snippets: List<Snippet>, query: String): List<Snippet> =
    filterSnippetItems(snippets, query, { it.cue }, { it.expansion })

fun <T> filterSnippetItems(
    snippets: List<T>,
    query: String,
    cue: (T) -> String,
    expansion: (T) -> String,
): List<T> {
    val needle = normalizeSearchText(query)
    if (needle.isEmpty()) return snippets
    return snippets.filter { normalizeSearchText("${cue(it)} ${expansion(it)}").contains(needle) }
}

private fun normalizeSearchText(value: String): String =
    value.lowercase().trim().replace(Regex("\\s+"), " ")

private fun normalizeCue(s: String): String =
    s.lowercase()
        .replace(Regex("\\s+"), " ")
        .trim()
        .trim('.', ',', '!', '?', ';', ':', '"', '\'')
        .trim()

File: src/shared/snippets.ts
// Voice snippets: speak a cue, paste a saved block. Pure matcher shared with the store layer;
// mirrors the Android snippet/Snippets.kt so a cue authored on the phone expands on the desktop.

export interface Snippet {
  cue: string
  expansion: string
}

export interface StoredSnippet extends Snippet {
  id: number
  created_at: number
}

/**
 * If `text` (the whole dictation) matches a snippet's cue — ignoring case, surrounding whitespace,
 * collapsed inner spaces, and trailing sentence punctuation — return its expansion; else null. The
 * whole-utterance match means saying a cue inside a sentence inserts the words literally. First wins.
 */
export function expandSnippet(text: string, snippets: Snippet[]): string | null {
  const key = normalizeCue(text)
  if (!key) return null
  const hit = snippets.find((s) => normalizeCue(s.cue) === key)
  return hit ? hit.expansion : null
}

function normalizeCue(s: string): string {
  return s
    .toLowerCase()
    .replace(/\s+/g, ' ')
    .trim()
    .replace(/^[.,!?;:'"]+|[.,!?;:'"]+$/g, '')
    .trim()
}
